Detectors/TOF/reconstruction/src/Decoder.cxx

```diff
@@ -0,0 +1,200 @@
+#include "TOFReconstruction/Decoder.h"
+#include <iostream>
+#include <chrono>
+#include <algorithm>
+#include "CommonConstants/LHCConstants.h"
+#define VERBOSE
+
+namespace o2
+{
+namespace tof
+{
+namespace compressed
+{
+
+bool Decoder::open(std::string name)
+{
+  if (mFile.is_open()) {
+    std::cout << "Warning: a file was already open, closing" << std::endl;
+    mFile.close();
+  }
+  mFile.open(name.c_str(), std::fstream::in | std::fstream::binary);
+  if (!mFile.is_open()) {
+    std::cerr << "Cannot open " << name << std::endl;
+    return true;
+  }
+
+  if (mBuffer)
+    delete[] mBuffer;
+
+  mFile.seekg(0, mFile.end);
+  mSize = mFile.tellg();
+  mFile.seekg(0);
+
+  mBuffer = new char[mSize];
+
+  // read content of infile
+  mFile.read(mBuffer, mSize);
+  mUnion = reinterpret_cast<Union_t*>(mBuffer);
+  mUnionEnd = reinterpret_cast<Union_t*>(mBuffer + mSize - 1);
+  close();
+
+  return false;
+}
+
+bool Decoder::close()
+{
+  if (mFile.is_open())
+    mFile.close();
+  return false;
+}
+
+void Decoder::readTRM(std::vector<Digit>* digits, int iddl, int orbit, int bunchid)
+{
+  if (mVerbose)
+    printTRMInfo();
+  int nhits = mUnion->frameHeader.numberOfHits;
+  int time_ext = mUnion->frameHeader.frameID << 13;
+  int itrm = mUnion->frameHeader.trmID;
+  mUnion++;
+
+  // read hits
+  Int_t channel, echannel;
+  Int_t tdc;
+  Int_t tot;
+  Int_t bc;
+  Int_t time;
+
+  Int_t digitInfo[4];
+
+  for (int i = 0; i < nhits; i++) {
+    FromRawHit2Digit(iddl, itrm, mUnion->packedHit.tdcID, mUnion->packedHit.chain, mUnion->packedHit.channel, orbit, bunchid, time_ext + mUnion->packedHit.time, mUnion->packedHit.tot, digitInfo);
+
+    if (mVerbose)
+      printHitInfo();
+    digits->emplace_back(digitInfo[0], digitInfo[1], digitInfo[2], digitInfo[3]);
+    mUnion++;
+  }
+}
+
+void Decoder::FromRawHit2Digit(int iddl, int itrm, int itdc, int ichain, int channel, int orbit, int bunchid, int tdc, int tot, int* digitInfo)
+{
+  // convert raw info in digit info (channel, tdc, tot, bc)
+  // tdc = packetHit.time + (frameHeader.frameID << 13)
+  int echannel = Geo::getECHFromIndexes(iddl, itrm, ichain, itdc, channel);
+  digitInfo[0] = Geo::getCHFromECH(echannel);
+  digitInfo[2] = tot;
+
+  digitInfo[3] = int(orbit * o2::tof::Geo::BC_IN_ORBIT);
+  digitInfo[3] += bunchid;
+  digitInfo[3] += tdc / 1024;
```

Comment:
this one also does not look to be correct: does bunchid + tdc/1024 makes sense?

Reply:
in the electronic map time is constructed from 3 components
orbit(when TOF window starts) + BC_IN_ORBIT(when TOF window starts) + TDCcounts_from_the_window_start

Having in mind that
1 orbit = 3564 BC
1 BC = 1024 TDC

In reconstruction we encode time in the digit
using BC(when the hit happened) + TDC_WITHIN_THE_BC(0:1023)

then when decoding I take out from TDCcounts_from_the_window_start
the number of BC from the stars of the tof window (added to BC) and I save in TDC only TDCcounts in the current BC

Reply:
sorry, I still don't understand what is the supposed range of 

1. bunchid + tdc / 1024;
2. line 88 leaves it only the room from 0 : N_BC_in_orbit

Reply:
The time will be derived as

time(in TDC channels) = BC(when the hit happened)*1024 + TDC_WITHIN_THE_BC(0:1023)
or
time(in ps) = BC(when the hit happened) * LHCBunchSpacingNS *1E3 + TDC_WITHIN_THE_BC(0:1023) * TDCBIN

with
static constexpr Float_t TDCBIN = o2::constants::lhc::LHCBunchSpacingNS*1E3/1024;  ///< TDC bin width [ps]

Reply:
Sorry, I still don't understand: you are using int and after the BC*3564 only the lowest 3564 are still free for extra info. How do you store there additionally the time with precision better than 1 BC?

Reply:
The next line (line 91)
  digitInfo[1] = tdc % 1024;

it is stored then in a different data member  --> mTDC(0:1023) in Digit class

Reply:
OK, then I probably don't understand what means BC(when the hit happened) and from which moment TDC counts. Anyway, we can discuss this later. Could you squash the commits  leaving only 2 for wokflow fixes and decoding? Or I can squash it before merging but the message will be taken from 1st one.

Reply:
I squashed leaving only 3 commits.